### ISSUE TITLE
failures when running multiple mpirun instances on more than 1 node in a cluster via qsub

### ISSUE THREAD

**Issue by Alisyed09** [open-mpi/ompi#1449]
Problem
We have a cluster of 28-core nodes using Moab/Torque where the following problem occurs with OpenMPI version 1.10.2 recently MPI was upgraded 1.8.X

When running multiple mpirun instances on more than 1 nodes mpirun calls fails. it can be run manually without any problems but with qsub it fails everytime. with the same error: 

There are not enough slots available in the system to satisfy the 8 slots 
that were requested by the application: 
/services/tools/gromacs/5.1.2/bin/mdrun_mpi 

For the latest submit, I added some flags for debugging to the command: 
mpirun -np 8 --map-by node --bind-to core --mca btl sm,self --display-allocation $cmd 

```
PBS_NODEFILE: /cm/local/apps/torque/var/spool/aux//2291493.r04-sn064
r06-cn207 
r06-cn206 
r06-cn205 
r06-cn204 
NPROCS: 4 
PBS_NP: 4 
Directory: 
PBS_O_WORKDIR: /home/people/erhh/tests/11_febr-19-13 

====================== ALLOCATED NODES ====================== 
r06-cn207: slots=1 max_slots=0 slots_inuse=0 state=UP 
r06-cn206: slots=1 max_slots=0 slots_inuse=0 state=UP 
r06-cn205: slots=1 max_slots=0 slots_inuse=0 state=UP 
r06-cn204: slots=1 max_slots=0 slots_inuse=0 state=UP 
================================================================= 
-------------------------------------------------------------------------- 
There are not enough slots available in the system to satisfy the 8 slots 
that were requested by the application: 
/services/tools/gromacs/5.1.2/bin/mdrun_mpi 

Either request fewer slots for your application, or make more slots available 
for use. 
-------------------------------------------------------------------------- 
```

It seems like OpeMPI is _not_ recognizing any of the 28 slots available on each of the nodes r06-cn204..207 

the PBS nodesfiles is in following format:

```
r06-cn204 np=28 thinnode compute switch-rack-06a infiniband
r06-cn205 np=28 thinnode compute switch-rack-06a infiniband
r06-cn206 np=28 thinnode compute switch-rack-06a infiniband
r06-cn207 np=28 thinnode compute switch-rack-06a infiniband
```

Versions tested
We tested mpirun (Open MPI) 1.10.2

Environment
The cluster is running Centos 7.2.

I hope this was understandable. I might have used some wrong terms....

**EDIT** Added verbatim blocks

**Comment by rhc54** [comment]
> On Mar 10, 2016, at 12:34 PM, Alisyed09 <notifications@github.com> wrote:
> 
> Problem
> We have a cluster of 28-core nodes using Moab/Torque where the following problem occurs with OpenMPI version 1.10.2 recently MPI was upgraded 1.8.X
> 
> When running multiple mpirun instances on more than 1 nodes mpirun calls fails. it can be run manually without any problems but with qsub it fails everytime. with the same error:
> 
> There are not enough slots available in the system to satisfy the 8 slots 
> that were requested by the application: 
> /services/tools/gromacs/5.1.2/bin/mdrun_mpi
> 
> For the latest submit, I added some flags for debugging to the command: 
> mpirun -np 8 --map-by node --bind-to core --mca btl sm,self --display-allocation $cmd
> 
The PBS_NODEFILE below doesn’t look right to me - it is supposed to list each node’s name once for each allocated slot on that node.
> PBS_NODEFILE: /cm/local/apps/torque/var/spool/aux//2291493.r04-sn064
> r06-cn207 
> r06-cn206 
> r06-cn205 
> r06-cn204 
> NPROCS: 4 
> PBS_NP: 4 
> Directory: 
> PBS_O_WORKDIR: /home/people/erhh/tests/11_febr-19-13
> 
> ====================== ALLOCATED NODES ====================== 
> r06-cn207: slots=1 max_slots=0 slots_inuse=0 state=UP 
> r06-cn206: slots=1 max_slots=0 slots_inuse=0 state=UP 
> r06-cn205: slots=1 max_slots=0 slots_inuse=0 state=UP
> 
> r06-cn204: slots=1 max_slots=0 slots_inuse=0 state=UP
> 
> There are not enough slots available in the system to satisfy the 8 slots 
> that were requested by the application: 
> /services/tools/gromacs/5.1.2/bin/mdrun_mpi
> 
> Either request fewer slots for your application, or make more slots available
> 
> for use.
> 
> It seems like OpeMPI is not recognizing any of the 28 slots available on each of the nodes r06-cn204..207
> 
> the PBS nodesfiles is in following format:
> 
> 

I have never seen the format below for a PBS_NODEFILE - where did it come from??

**Comment by Alisyed09** [comment]
Dear rhc54,

the nodes file is pbs_server nodes file under server_priv and we have been using this for last year. did i understood your question right ?

i did suspect your below comment as well. is there a way i can verify this ?

> The PBS_NODEFILE below doesn’t look right to me - it is supposed to list each node’s name once for each allocated slot on that node.
 PBS_NODEFILE: /cm/local/apps/torque/var/spool/aux//2291493.r04-sn064
 r06-cn207
 r06-cn206
 r06-cn205
 r06-cn204
 NPROCS: 4
 PBS_NP: 4
 Directory:
 PBS_O_WORKDIR: /home/people/erhh/tests/11_febr-19-13

>

**Comment by rhc54** [comment]
> On Mar 10, 2016, at 1:43 PM, Alisyed09 <notifications@github.com> wrote:
> 
> Dear rhc54,
> 
> the nodes file is pbs_server nodes file under server_priv and we have been using this for last year. did i understood your question right ?
> 
> 

Ah, okay - that’s the Torque configuration file, not the file that contains the allocation info for this job.
> i did suspect your below comment as well. is there a way i can verify this ?
> 
> The PBS_NODEFILE below doesn’t look right to me - it is supposed to list each node’s name once for each allocated slot on that node.
> PBS_NODEFILE: /cm/local/apps/torque/var/spool/aux//2291493.r04-sn064
> r06-cn207
> r06-cn206
> r06-cn205
> r06-cn204
> NPROCS: 4
> PBS_NP: 4
> Directory:
> PBS_O_WORKDIR: /home/people/erhh/tests/11_febr-19-13
> 
> 
It looks to me like the PBS_NODEFILE format may have changed. What version of Moab are you running?

**Comment by ggouaillardet** [comment]
@Alisyed09 ```pbs_server_nodes``` list all the nodes of the cluster.
within a job, mpirun will only use the resources allocated by PBS, and you can see that by running
```
cat $PBS_NODEFILE
```
from within your job.

the logs clearly state 4 slots have been allocated, but you try to run 8 tasks.
if this is what you really want to do, you need to
```mpirun --oversubscribe```
but it is more likely you want to allocate 8 slots
(you can do that by updating your ```qsub``` command line or the PBS sentinels at the beginning of your script)

**Comment by rhc54** [comment]
I think the issue here is that the PBS_NODEFILE format is telling us that we only have one slot allocated per node. @Alisyed09 can you please post the qsub command you used to get this allocation?

My best guess is that Moab has been configured to support sharing of nodes between users, and thus does not allocate the entire node to the user by default. Thus, you must indicate in your qsub request that you want all the slots on each allocated node. Otherwise, Moab is only allocating you one slot on each of four nodes.

**Comment by Alisyed09** [comment]
Dear Both,

Thanks for the assistance, Appreciated!  

This is how my submit script looks like:

module load openmpi/gcc/64/1.10.2
module load gromacs/5.1.2

WORKDIR=$PBS_O_WORKDIR
cd $WORKDIR
cp $PBS_NODEFILE MY_NODEFILE
NPROCS=`wc -l < $PBS_NODEFILE`

echo 'Number of procs asked is'
echo "PBS_NODEFILE: $PBS_NODEFILE"
cat $PBS_NODEFILE
echo "NPROCS: $NPROCS"
echo "PBS_NP: $PBS_NP"
echo 'Directory:'
echo "PBS_O_WORKDIR: $PBS_O_WORKDIR"

mdrun=$(which mdrun_mpi)
cmd="$mdrun -s topol -cpi -v -maxh 8"
mpirun -np 8 --map-by socket:pe=8 --display-allocation $cmd

I submit it using:

qsub script name  and no parametters

output 

Number of procs asked is
PBS_NODEFILE: /cm/local/apps/torque/var/spool/aux//2293982.r04-sn064.cm.cluster
r06-cn207
r06-cn206
r06-cn205
r06-cn204
NPROCS: 4
PBS_NP: 4
Directory:
PBS_O_WORKDIR: /home/people/erhh/tests/11_febr-19-13

======================   ALLOCATED NODES   ======================
	r06-cn207: slots=1 max_slots=0 slots_inuse=0 state=UP
	r06-cn206: slots=1 max_slots=0 slots_inuse=0 state=UP
	r06-cn205: slots=1 max_slots=0 slots_inuse=0 state=UP
	r06-cn204: slots=1 max_slots=0 slots_inuse=0 state=UP
=================================================================
--------------------------------------------------------------------------
There are not enough slots available in the system to satisfy the 8 slots 
that were requested by the application:
  /services/tools/gromacs/5.1.2/bin/mdrun_mpi

Either request fewer slots for your application, or make more slots available
for use.
--------------------------------------------------------------------------


NODEFILE

r06-cn207
r06-cn206
r06-cn205
r06-cn204


Thanks again for your help !

**Comment by ggouaillardet** [comment]
so basically you do not tell pbs how many slots you want.
if you want 8 slots on one node, you can try
qsub -l nodes=1:ppn=8 ...
but if you really want pbs decide how many slots should be allocated, you have to
mpirun --oversubscribe ...
note that since you request the mapping in your script, you might end up having several jobs using the same resources, when other available resources are left unused.
bottom line, the safest option is probably to explicitly request pbs how many resources should be allocated, and then simply
mpirun $cmd
in your script

**Comment by Alisyed09** [comment]
I tried submitting a job as you suggest

qsub -l nodes=4:ppn=28 submit.sh

the job failed with this output:


Number of procs asked is
PBS_NODEFILE: /cm/local/apps/torque/var/spool/aux//2294617.r04-sn064.cm.cluster
r06-cn207
r06-cn206
r06-cn205
r06-cn204
NPROCS: 4
PBS_NP: 4
Directory:
PBS_O_WORKDIR: /home/people/erhh/tests/11_febr-19-13

======================   ALLOCATED NODES   ======================
	r06-cn207: slots=1 max_slots=0 slots_inuse=0 state=UP
	r06-cn206: slots=1 max_slots=0 slots_inuse=0 state=UP
	r06-cn205: slots=1 max_slots=0 slots_inuse=0 state=UP
	r06-cn204: slots=1 max_slots=0 slots_inuse=0 state=UP
=================================================================
--------------------------------------------------------------------------
There are not enough slots available in the system to satisfy the 8 slots
that were requested by the application:
  /services/tools/gromacs/5.1.2/bin/mdrun_mpi

Either request fewer slots for your application, or make more slots available
for use.
--------------------------------------------------------------------------

does this ring any bells ?

Job script is as below:

#!/bin/bash
#PBS -V
#PBS -N tmp
#PBS -W group_list=pr_hackinars
#PBS -A pr_hackinars
###PBS -l procs=30
###PBS -l procs=28
###PBS -l nodes=1:ppn=28
#PBS -l nodes=4:ppn=28
#PBS -q batch
#PBS -l walltime=8:00:00
#PBS -W x=FLAGS:ADVRES:erhh.35058

# OpenMPI
module load openmpi/gcc/64/1.10.2
module load gromacs/5.1.2
WORKDIR=$PBS_O_WORKDIR
cd $WORKDIR
cp $PBS_NODEFILE MY_NODEFILE
NPROCS=`wc -l < $PBS_NODEFILE`


echo 'Number of procs asked is'
echo "PBS_NODEFILE: $PBS_NODEFILE"
cat $PBS_NODEFILE
echo "NPROCS: $NPROCS"
echo "PBS_NP: $PBS_NP"
echo 'Directory:'
echo "PBS_O_WORKDIR: $PBS_O_WORKDIR"

mdrun=$(which mdrun_mpi)
cmd="$mdrun -s topol -cpi -v -maxh 8"
mpirun -np 8 --map-by socket:pe=8 --display-allocation $cmd

**Comment by jsquyres** [comment]
@Alisyed09 Github pro tip: please [enclose your blocks of verbatim text in triple-single-ticks](https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code) so that they are rendered properly in the wiki (more specifically: so that that verbatim text doesn't accidentally get rendered as a large heading or some other accidental wiki syntax).  Thanks!

**Comment by Alisyed09** [comment]
ah thanks for the tip you can clearly see a github newbie :+1:

**Comment by rhc54** [comment]
@Alisyed09 Can you tell us what version of Moab you are using? It looks like the PBS_NODEFILE format has changed, and we'll probably need to revise our parsing algorithm.

**Comment by Alisyed09** [comment]
Hi rhc54,

we upgraded from  4.2.6.1 pbs_mom to  Version: 5.1.1.h1 and job seems to run.

I will post the results once the job finished.

**Comment by ggouaillardet** [comment]
@Alisyed09 that sounds weird to me ... it looks like ppn=28 was simply ignored
before invoking mpirun, can you please insert
```
env
```
in your script and post the output ?

from https://github.com/open-mpi/ompi/blob/master/orte/mca/ras/tm/ras_tm_component.c

```c
    /* for big SMP machines (e.g., those from SGI), listing the nodes
     * once/slot in the nodefile is extreme. In those cases, they may
     * choose to list each node once, but then provide an envar that
     * tells us how many cpus/node were allocated. Allow the user to
     * inform us that we are in such an environment
     */
    mca_ras_tm_component.smp_mode = false;
    (void) mca_base_component_var_register (c, "smp",
                                            "The Torque system is configured in SMP mode "
                                            "with the number of cpus/node given in the environment",
                                            MCA_BASE_VAR_TYPE_BOOL, NULL, 0, 0,
                                            OPAL_INFO_LVL_9,
                                            MCA_BASE_VAR_SCOPE_READONLY,
                                            &mca_ras_tm_component.smp_mode);
```
you can also try

```
mpirun --mca ras_tm_smp 1 ...
```

btw, was ompi built with tm ?
if you configure'd with ```--with-tm``` then you should be fine
otherwise, you can
```ompi_info --all``` and grep tm

**Comment by Alisyed09** [comment]
Hi All, A quick update testing during weekend shows the PBS_mom seems to have fixed the issue. i will keep you posted on the scaling of the open mpi but we can close the thread on this note it was 4.2.61 or lower pbs_mom will have parsing problem with PBS_NODEFILE.
